{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip3 install git+https://github.com/MaayanLab/blitzgsea.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'LIBRARY_LIST_URL': 'https://maayanlab.cloud/speedrichr/api/listlibs', 'LIBRARY_DOWNLOAD_URL': 'https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName='}\n",
      "0 - GeneSigDB\n",
      "1 - Enrichr_Submissions_TF-Gene_Coocurrence\n",
      "2 - SysMyo_Muscle_Gene_Sets\n",
      "3 - WikiPathway_2021_Human\n",
      "4 - HomoloGene\n",
      "5 - WikiPathways_2013\n",
      "6 - OMIM_Disease\n",
      "7 - Data_Acquisition_Method_Most_Popular_Genes\n",
      "8 - Cancer_Cell_Line_Encyclopedia\n",
      "9 - WikiPathways_2016\n",
      "10 - WikiPathways_2015\n",
      "11 - RNAseq_Automatic_GEO_Signatures_Human_Up\n",
      "12 - Human_Gene_Atlas\n",
      "13 - Kinase_Perturbations_from_GEO_down\n",
      "14 - Disease_Signatures_from_GEO_down_2014\n",
      "15 - Disease_Perturbations_from_GEO_up\n",
      "16 - Old_CMAP_down\n",
      "17 - MCF7_Perturbations_from_GEO_up\n",
      "18 - NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions\n",
      "19 - DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019\n",
      "20 - PPI_Hub_Proteins\n",
      "21 - Disease_Signatures_from_GEO_up_2014\n",
      "22 - GTEx_Tissue_Expression_Up\n",
      "23 - L1000_Kinase_and_GPCR_Perturbations_up\n",
      "24 - ARCHS4_Cell-lines\n",
      "25 - VirusMINT\n",
      "26 - KEGG_2019_Human\n",
      "27 - ARCHS4_Tissues\n",
      "28 - MGI_Mammalian_Phenotype_Level_4\n",
      "29 - MGI_Mammalian_Phenotype_Level_3\n",
      "30 - InterPro_Domains_2019\n",
      "31 - KEGG_2015\n",
      "32 - MSigDB_Computational\n",
      "33 - TF-LOF_Expression_from_GEO\n",
      "34 - KEGG_2013\n",
      "35 - GWAS_Catalog_2019\n",
      "36 - KEGG_2016\n",
      "37 - NCI-Nature_2015\n",
      "38 - NCI-Nature_2016\n",
      "39 - CCLE_Proteomics_2020\n",
      "40 - PheWeb_2019\n",
      "41 - RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\n",
      "42 - LINCS_L1000_Chem_Pert_down\n",
      "43 - Old_CMAP_up\n",
      "44 - LINCS_L1000_Ligand_Perturbations_down\n",
      "45 - Enrichr_Users_Contributed_Lists_2020\n",
      "46 - NIH_Funded_PIs_2017_Human_GeneRIF\n",
      "47 - Jensen_TISSUES\n",
      "48 - Azimuth_Cell_Types_2021\n",
      "49 - Panther_2016\n",
      "50 - DisGeNET\n",
      "51 - LINCS_L1000_Ligand_Perturbations_up\n",
      "52 - Rare_Diseases_AutoRIF_Gene_Lists\n",
      "53 - TargetScan_microRNA\n",
      "54 - Achilles_fitness_increase\n",
      "55 - Panther_2015\n",
      "56 - WikiPathways_2019_Mouse\n",
      "57 - ARCHS4_TFs_Coexp\n",
      "58 - LINCS_L1000_Chem_Pert_up\n",
      "59 - MSigDB_Oncogenic_Signatures\n",
      "60 - Gene_Perturbations_from_GEO_down\n",
      "61 - Table_Mining_of_CRISPR_Studies\n",
      "62 - Rare_Diseases_GeneRIF_Gene_Lists\n",
      "63 - Ligand_Perturbations_from_GEO_down\n",
      "64 - SILAC_Phosphoproteomics\n",
      "65 - Ligand_Perturbations_from_GEO_up\n",
      "66 - Drug_Perturbations_from_GEO_up\n",
      "67 - MGI_Mammalian_Phenotype_Level_4_2021\n",
      "68 - Allen_Brain_Atlas_10x_scRNA_2021\n",
      "69 - ClinVar_2019\n",
      "70 - KEA_2015\n",
      "71 - KEA_2013\n",
      "72 - Chromosome_Location\n",
      "73 - Microbe_Perturbations_from_GEO_up\n",
      "74 - COVID-19_Related_Gene_Sets_2021\n",
      "75 - MGI_Mammalian_Phenotype_Level_4_2019\n",
      "76 - ARCHS4_IDG_Coexp\n",
      "77 - NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions\n",
      "78 - lncHUB_lncRNA_Co-Expression\n",
      "79 - DrugMatrix\n",
      "80 - Virus_Perturbations_from_GEO_down\n",
      "81 - huMAP\n",
      "82 - L1000_Kinase_and_GPCR_Perturbations_down\n",
      "83 - Elsevier_Pathway_Collection\n",
      "84 - NIH_Funded_PIs_2017_Human_AutoRIF\n",
      "85 - RNAseq_Automatic_GEO_Signatures_Mouse_Down\n",
      "86 - ESCAPE\n",
      "87 - UK_Biobank_GWAS_v1\n",
      "88 - Aging_Perturbations_from_GEO_up\n",
      "89 - Human_Phenotype_Ontology\n",
      "90 - dbGaP\n",
      "91 - SubCell_BarCode\n",
      "92 - Transcription_Factor_PPIs\n",
      "93 - GO_Cellular_Component_2017b\n",
      "94 - MSigDB_Hallmark_2020\n",
      "95 - BioPlanet_2019\n",
      "96 - HDSigDB_Human_2021\n",
      "97 - GTEx_Tissue_Expression_Down\n",
      "98 - Tissue_Protein_Expression_from_Human_Proteome_Map\n",
      "99 - Epigenomics_Roadmap_HM_ChIP-seq\n",
      "100 - PhenGenI_Association_2021\n",
      "101 - MCF7_Perturbations_from_GEO_down\n",
      "102 - ProteomicsDB_2020\n",
      "103 - Virus-Host_PPI_P-HIPSTer_2020\n",
      "104 - OMIM_Expanded\n",
      "105 - Genes_Associated_with_NIH_Grants\n",
      "106 - CellMarker_Augmented_2021\n",
      "107 - ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X\n",
      "108 - HDSigDB_Mouse_2021\n",
      "109 - Jensen_COMPARTMENTS\n",
      "110 - ChEA_2015\n",
      "111 - ChEA_2016\n",
      "112 - KEGG_2019_Mouse\n",
      "113 - Drug_Perturbations_from_GEO_2014\n",
      "114 - TargetScan_microRNA_2017\n",
      "115 - KEGG_2021_Mouse\n",
      "116 - Allen_Brain_Atlas_down\n",
      "117 - BioCarta_2013\n",
      "118 - Reactome_2013\n",
      "119 - WikiPathways_2019_Human\n",
      "120 - Gene_Perturbations_from_GEO_up\n",
      "121 - GO_Cellular_Component_2015\n",
      "122 - GO_Cellular_Component_2013\n",
      "123 - BioCarta_2016\n",
      "124 - Reactome_2015\n",
      "125 - BioCarta_2015\n",
      "126 - GO_Cellular_Component_2018\n",
      "127 - Reactome_2016\n",
      "128 - GO_Cellular_Component_2017\n",
      "129 - GO_Cellular_Component_2021\n",
      "130 - WikiPathway_2021_Mouse\n",
      "131 - ENCODE_TF_ChIP-seq_2015\n",
      "132 - ENCODE_TF_ChIP-seq_2014\n",
      "133 - RNAseq_Automatic_GEO_Signatures_Mouse_Up\n",
      "134 - GO_Molecular_Function_2017b\n",
      "135 - MGI_Mammalian_Phenotype_2013\n",
      "136 - HMS_LINCS_KinomeScan\n",
      "137 - NCI-60_Cancer_Cell_Lines\n",
      "138 - MGI_Mammalian_Phenotype_2017\n",
      "139 - Rare_Diseases_GeneRIF_ARCHS4_Predictions\n",
      "140 - Virus_Perturbations_from_GEO_up\n",
      "141 - Enrichr_Libraries_Most_Popular_Genes\n",
      "142 - Orphanet_Augmented_2021\n",
      "143 - GO_Biological_Process_2021\n",
      "144 - TRANSFAC_and_JASPAR_PWMs\n",
      "145 - Rare_Diseases_AutoRIF_ARCHS4_Predictions\n",
      "146 - COVID-19_Related_Gene_Sets\n",
      "147 - Kinase_Perturbations_from_GEO_up\n",
      "148 - Descartes_Cell_Types_and_Tissue_2021\n",
      "149 - TF_Perturbations_Followed_by_Expression\n",
      "150 - GTEx_Aging_Signatures_2021\n",
      "151 - Tissue_Protein_Expression_from_ProteomicsDB\n",
      "152 - Aging_Perturbations_from_GEO_down\n",
      "153 - GO_Biological_Process_2013\n",
      "154 - GO_Biological_Process_2017b\n",
      "155 - GO_Biological_Process_2018\n",
      "156 - CORUM\n",
      "157 - Phosphatase_Substrates_from_DEPOD\n",
      "158 - BioPlex_2017\n",
      "159 - GO_Biological_Process_2015\n",
      "160 - TRRUST_Transcription_Factors_2019\n",
      "161 - GO_Biological_Process_2017\n",
      "162 - Pfam_InterPro_Domains\n",
      "163 - HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression\n",
      "164 - Pfam_Domains_2019\n",
      "165 - Allen_Brain_Atlas_up\n",
      "166 - NURSA_Human_Endogenous_Complexome\n",
      "167 - Genome_Browser_PWMs\n",
      "168 - HumanCyc_2015\n",
      "169 - HumanCyc_2016\n",
      "170 - Chromosome_Location_hg19\n",
      "171 - ChEA_2013\n",
      "172 - Mouse_Gene_Atlas\n",
      "173 - miRTarBase_2017\n",
      "174 - Jensen_DISEASES\n",
      "175 - RNAseq_Automatic_GEO_Signatures_Human_Down\n",
      "176 - ARCHS4_Kinases_Coexp\n",
      "177 - Microbe_Perturbations_from_GEO_down\n",
      "178 - PanglaoDB_Augmented_2021\n",
      "179 - ENCODE_Histone_Modifications_2013\n",
      "180 - Achilles_fitness_decrease\n",
      "181 - ENCODE_Histone_Modifications_2015\n",
      "182 - DSigDB\n",
      "183 - Disease_Perturbations_from_GEO_down\n",
      "184 - DepMap_WG_CRISPR_Screens_Broad_CellLines_2019\n",
      "185 - Drug_Perturbations_from_GEO_down\n",
      "186 - GO_Molecular_Function_2021\n",
      "187 - GO_Molecular_Function_2017\n",
      "188 - GO_Molecular_Function_2018\n",
      "189 - GO_Molecular_Function_2013\n",
      "190 - GO_Molecular_Function_2015\n",
      "191 - TG_GATES_2020\n",
      "192 - KEGG_2021_Human\n",
      "193 - HMDB_Metabolites\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calibration: 100%|██████████| 25/25 [00:35<00:00,  1.41s/it]\n",
      "Enrichment : 100%|██████████| 320/320 [00:00<00:00, 569.67it/s]\n"
     ]
    }
   ],
   "source": [
    "import blitzgsea as blitz\n",
    "import urllib.request\n",
    "import pandas as pd\n",
    "\n",
    "# download example gene expression signature\n",
    "url = \"https://github.com/MaayanLab/blitzgsea/raw/main/testing/ageing_muscle_gtex.tsv\"\n",
    "urllib.request.urlretrieve(url, \"ageing_muscle_gtex.tsv\")\n",
    "\n",
    "# read signature as pandas dataframe\n",
    "signature = pd.read_csv(\"ageing_muscle_gtex.tsv\")\n",
    "\n",
    "# list available gene set libraries in Enrichr\n",
    "#blitz.enrichr.print_libraries()\n",
    "\n",
    "# use enrichr submodule to retrieve gene set library\n",
    "library = blitz.enrichr.get_library(\"KEGG_2021_Human\")\n",
    "\n",
    "# run enrichment analysis\n",
    "result = blitz.gsea(signature, library)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "import prismx as px\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "px.bridge_gsea()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "60f3b3101667592c150dcbaf93ef24f0a05e4dd4dc38244d9b99f2435a60626a"
  },
  "kernelspec": {
   "display_name": "Python 3.8.5 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
